File: benchmarks/alpaca/src/MathParser.scala
package bench.alpaca

import alpaca.*

val MathLexer = lexer {
  case "\\s+" => Token.Ignored
  case x @ ("\\+" | "-" | "\\*" | "/" | "\\(" | "\\)") => Token[x.type]
  case num @ "\\d+" => Token["num"](num.toInt)
}

object MathParser extends Parser:
  val Expr: Rule[Int] = rule(
    "mul" { case (Expr(a), MathLexer.`\\*`(_), Expr(b)) => a * b },
    "div" { case (Expr(a), MathLexer.`/`(_), Expr(b)) => a / b },
    "plus" { case (Expr(a), MathLexer.`\\+`(_), Expr(b)) => a + b },
    "minus" { case (Expr(a), MathLexer.`-`(_), Expr(b)) => a - b },
    { case (MathLexer.`\\(`(_), Expr(a), MathLexer.`\\)`(_)) => a },
    { case MathLexer.num(n) => n.value },
  )

  val root: Rule[Int] = rule { case Expr(v) => v }

  override val resolutions = Set(
    // multiplication and division are left associative (reduction before shift)
    production.mul.before(MathLexer.`\\*`, MathLexer.`/`),
    production.div.before(MathLexer.`\\*`, MathLexer.`/`),
    // addition and subtraction have lower precedence than multiplication and division
    production.plus.after(MathLexer.`\\*`, MathLexer.`/`),
    production.minus.after(MathLexer.`\\*`, MathLexer.`/`),
    // addition and subtraction are left associative (reduction before shift)
    production.plus.before(MathLexer.`\\+`, MathLexer.`-`),
    production.minus.before(MathLexer.`\\+`, MathLexer.`-`),
  )

@main def mathParserMain(): Unit =
  import java.nio.file.{Files, Paths}

  val filePathIterative = s"inputs/iterative_math_3.txt"
  val fileContentIterative = new String(
    Files.readAllBytes(Paths.get(filePathIterative)),
  )

  try
    val (_, tokens) = MathLexer.tokenize(fileContentIterative)
    val (_, result) = MathParser.parse(tokens)
    println(s"\nResult Iterative: $result")
  catch
    case e: Exception =>
      println(s"\nError Iterative: ${e.getMessage}")
      e.printStackTrace()

  val filePathRecursive = s"inputs/recursive_math_3.txt"
  val fileContentRecursive = new String(
    Files.readAllBytes(Paths.get(filePathRecursive)),
  )

  try
    val (_, tokens) = MathLexer.tokenize(fileContentRecursive)
    val (_, result) = MathParser.parse(tokens)
    println(s"\nResult Recursive: $result")
  catch
    case e: Exception =>
      println(s"\nError Recursive: ${e.getMessage}")
      e.printStackTrace()

File: benchmarks/sly/math_parser.py
from pathlib import Path
from sly import Lexer, Parser

class MathLexer(Lexer):
    literals = ['+', '-', '*', '/', '(', ')']
    tokens = {NUM}
    ignore = ' \t\n'

    NUM = r'\d+'
    
    def error(self, t):
        print(f"Illegal character '{t.value[0]}'")
        self.index += 1


class MathParser(Parser):
    tokens = MathLexer.tokens

    precedence = (
        ('left', "+", "-"),
        ('left', "*", "/"),
    )

    @_('expr')
    def program(self, p):
        return p.expr

    @_('expr "*" expr')
    def expr(self, p):
        return p.expr0 * p.expr1

    @_('expr "/" expr')
    def expr(self, p):
        return p.expr0 / p.expr1

    @_('expr "+" expr')
    def expr(self, p):
        return p.expr0 + p.expr1
    
    @_('expr "-" expr')
    def expr(self, p):
        return p.expr0 - p.expr1
    
    @_('"(" expr ")"')
    def expr(self, p):
        return p.expr

    @_('NUM')
    def expr(self, p):
        return int(p.NUM)

    def error(self, p):
        if p:
            print(f"Syntax error at token {p.type} ('{p.value}')")
        else:
            print("Syntax error at EOF")


if __name__ == "__main__":
    # Test the parser
    lexer = MathLexer()
    parser = MathParser()

    file_path = Path(__file__).parent.parent / "inputs" / "iterative_math_3.txt"
    with file_path.open('r') as f:
        file_content = f.read()
    try:
        tokens = lexer.tokenize(file_content)
        result = parser.parse(tokens)
        print(f"Result Iterative: {result}")
    except Exception as e:
        print(f"Error Iterative: {e}")


    file_path = Path(__file__).parent.parent / "inputs" / "recursive_math_3.txt"
    with file_path.open('r') as f:
        file_content = f.read()
    try:
        tokens = lexer.tokenize(file_content)
        result = parser.parse(tokens)
        print(f"Result Recursive: {result}")
    except Exception as e:
        print(f"Error Recursive: {e}")
